File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        for (int[] row : matrix){
            if (row[0] <= target && target <= row[row.length-1]){
                if (search(row, target)) return true;
            }
        }
        return false;
    }
    
    public boolean search(int[] arr, int k){
        int l = 0;
        int h = arr.length - 1;
        
        while (l <= h){
            int m = (l+h) / 2;
            
            if (arr[m] == k) return true;
            else if (arr[m] > k){
                h = m - 1;
            }
            else {
                l = m + 1;
            }
        }
        return false;
    }
}

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.py
class Solution:  
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        def search(arr, k):
            l, h = 0, len(arr)-1
            while l <= h:
                m = (l+h) // 2
                if arr[m] == k:
                    return True
                elif arr[m] < k:
                    l = m + 1
                elif arr[m] > k:
                    h = m - 1
            return False
                    
        for row in matrix:
            if row and row[0] <= target <= row[-1]:
                if search(row, target): return True
        
        return False